controllers/etcdcopybackupstask/reconciler.go

```diff
@@ -283,6 +277,218 @@ func getConditionType(jobConditionType batchv1.JobConditionType) druidv1alpha1.C
+	}
+
+	volumes, err := getVolumes(&sourceStore, sourceProvider, "source-")
+	if err != nil {
+		return nil, err
+	}
+	job.Spec.Template.Spec.Volumes = append(job.Spec.Template.Spec.Volumes, volumes...)
+
+	volumes, err = getVolumes(&targetStore, targetProvider, "target-")
+	if err != nil {
+		return nil, err
+	}
+	job.Spec.Template.Spec.Volumes = append(job.Spec.Template.Spec.Volumes, volumes...)
+
+	if err := controllerutil.SetControllerReference(task, job, r.Scheme()); err != nil {
+		return nil, fmt.Errorf("could not set owner reference for job %v: %w", kutil.ObjectName(job), err)
+	}
+	return job, nil
+}
+
+func getVolumeNamePrefix(prefix string) string {
+	switch prefix {
+	case "source-":
+		return prefix
+	default:
+		return ""
+	}
+}
+
+func getVolumes(store *druidv1alpha1.StoreSpec, provider, prefix string) (volumes []corev1.Volume, err error) {
+	switch provider {
+	case druidutils.Local:
+		hostPathDirectory := corev1.HostPathDirectory
+		volumes = append(volumes, corev1.Volume{
+			Name: prefix + "host-storage",
+			VolumeSource: corev1.VolumeSource{
+				HostPath: &corev1.HostPathVolumeSource{
+					Path: *store.Container,
+					Type: &hostPathDirectory,
+				},
+			},
+		})
+	case druidutils.GCS, druidutils.S3, druidutils.ABS, druidutils.Swift, druidutils.OCS, druidutils.OSS:
+		if store.SecretRef == nil {
+			err = fmt.Errorf("no secretRef is configured for backup %sstore", prefix)
+			return
+		}
+		volumes = append(volumes, corev1.Volume{
+			Name: getVolumeNamePrefix(prefix) + "etcd-backup",
+			VolumeSource: corev1.VolumeSource{
+				Secret: &corev1.SecretVolumeSource{
+					SecretName: store.SecretRef.Name,
+				},
+			},
+		})
+
+	}
+	return
+}
+
+func getVolumeMounts(store *druidv1alpha1.StoreSpec, provider, volumeMountPrefix string) (volumeMounts []corev1.VolumeMount) {
+	switch provider {
+	case druidutils.Local:
+		volumeMounts = append(volumeMounts, corev1.VolumeMount{
+			Name:      volumeMountPrefix + "host-storage",
+			MountPath: *store.Container,
+		})
+	case druidutils.GCS:
+		volumeMounts = append(volumeMounts, corev1.VolumeMount{
+			Name:      getVolumeNamePrefix(volumeMountPrefix) + "etcd-backup",
+			MountPath: "/root/." + getVolumeNamePrefix(volumeMountPrefix) + "gcp/",
+		})
+	case druidutils.S3, druidutils.ABS, druidutils.Swift, druidutils.OCS, druidutils.OSS:
+		volumeMounts = append(volumeMounts, corev1.VolumeMount{
+			Name:      getVolumeNamePrefix(volumeMountPrefix) + "etcd-backup",
+			MountPath: "/root/" + getVolumeNamePrefix(volumeMountPrefix) + "etcd-backup/",
+		})
+	}
+	return
+}
+
+func getEnvVarFromValues(name, value string) corev1.EnvVar {
+	return corev1.EnvVar{
+		Name:  name,
+		Value: value,
+	}
+}
+
+func getProviderEnvs(store *druidv1alpha1.StoreSpec, storeProvider, prefix, volumePrefix string) []corev1.EnvVar {
+	env := []corev1.EnvVar{getEnvVarFromValues(prefix+"STORAGE_CONTAINER", *store.Container)}
```

Comment:
Proper way is to ensure that you have a fixed sized slice as you will at most have 2 elements in the slice.
```go
	envVars := make([]corev1.EnvVar, 2)
```